tests/e2e/e2e/executor/executor.go

```diff
@@ -0,0 +1,143 @@
+}
+
+// Cleanuper is an interface that defines a cleanup operation.
+// A step should implement this interface if it requires cleanup after execution.
+type Cleanuper interface {
+	// Cleanup performs the cleanup operation for the step.
+	Cleanup(*testing.T, client.Client) error
+}
+
+// StepsStack is a structure that holds steps in a stack-like manner.
+// It allows pushing new steps onto the stack and popping them off.
+// This is useful for managing the order of cleanup.
+type StepsStack struct {
+	steps []Step
+}
+
+func NewStepsStack() *StepsStack {
+	return &StepsStack{
+		steps: make([]Step, 0),
+	}
+}
+
+// Push adds a new step to the top of the stack.
+func (q *StepsStack) Push(step Step) {
+	q.steps = append(q.steps, step)
+}
+
+// Pop removes and returns the step at the top of the stack.
+func (q *StepsStack) Pop() Step {
+	if len(q.steps) == 0 {
+		return nil
+	}
+	step := q.steps[len(q.steps)-1]
+	q.steps = q.steps[:len(q.steps)-1]
+	return step
+}
+
+type Executor struct {
+	Options
+
+	t     *testing.T
+	steps *StepsStack
+
+	K8SClient client.Client
+}
+
+type Options struct {
+	IsCi        bool
+	OnlyCleanup bool
+}
+
+func NewExecutorWithOptionsFromEnv(t *testing.T) *Executor {
+	options := Options{
+		IsCi:        os.Getenv("CI") == "true",
+		OnlyCleanup: os.Getenv("ONLY_CLEANUP") == "true",
+	}
+
+	return NewExecutor(t, options)
+}
+
+func NewExecutor(t *testing.T, options Options) *Executor {
+	k8sClient, err := setup.ClientFromKubeconfig(t)
+	if err != nil {
+		t.Fatalf("Failed to create Kubernetes client: %s", err.Error())
+	}
+
+	return &Executor{
+		t:         t,
+		K8SClient: k8sClient,
+		Options:   options,
+		steps:     NewStepsStack(),
+	}
+}
+
+// RunStep executes a step and adds it to the executor's cleanup stack.
+func (e *Executor) RunStep(step Step) error {
+	e.steps.Push(step)
+	if e.OnlyCleanup {
+		logging.Debugf(e.t, "Skipping step execution in cleanup mode: %s", step.Description())
+		return nil
+	}
+
+	logging.Tracef(e.t, step.Description())
+
+	logging.Debugf(e.t, "Executing step: %s", step.Description())
+	if err := step.Execute(e.t, e.K8SClient); err != nil {
+		logging.Errorf(e.t, "Failed to execute step: %s err=%s", step.Description(), err.Error())
+		return err
+	}
+
+	logging.Untracef(e.t, step.Description())
+	return nil
+}
+
+// Cleanup performs the cleanup of all already executed steps in reverse order.
+// If the executor is running in a CI environment (Options.IsCi is true), cleanup is skipped.
+// If the executor is in cleanup mode (Options.OnlyCleanup is true),
+// it will only perform cleanup without executing any steps.
+func (e *Executor) Cleanup() {
+	if e.IsCi && !e.OnlyCleanup {
+		logging.Infof(e.t, "Skipping cleanup in CI environment")
+		return
+	}
+
+	logging.Tracef(e.t, "Starting cleanup")
+	defer logging.Untracef(e.t, "Finished cleanup")
+
+	// Perform cleanup in reverse order
+	for step := e.steps.Pop(); step != nil; step = e.steps.Pop() {
```

Comment:
Implement `e.steps.Empty()` method and use it as below. See: https://go.dev/play/p/9FXwxp5n4-h
```go
for !e.steps.Empty() {
```